src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs

```diff
@@ -85,12 +86,30 @@ internal class KafkaConsumerActor<K, V> : ActorBase, ILogReceive, IWithTimers
         private IImmutableSet<TopicPartitionOffset>
             _rebalanceCommitStash = ImmutableHashSet<TopicPartitionOffset>.Empty;
 
+        /// <summary>
+        /// Some important behavioral difference between the .NET and Java Kafka SDKs:
+        ///
+        /// 1.  Pausing partitions does not guarantee that previously retrieved messages stored
+        ///     inside <see cref="_consumer"/>'s buffer won't be returned.
+        /// 2.  The .NET SDK works from single messages rather than batches, therefore it is totally
+        ///     possible that assignment / re-balance events may occur in the middle of gathering messages
+        ///     via single Consume(int) calls.
+        /// 3. Therefore, it's possible that we will receive messages from partitions that have been
+        ///     assigned but not requested. Therefore, we need to make sure we buffer these messages until
+        ///     the <see cref="KafkaConsumerActorMetadata.Internal.RequestMessages"/> comes in.
+        ///
+        /// That's what this data structure is for - effectively it's a small stash of messages that have
+        /// been sent but not requested. This is designed to help us prevent issues such as
+        /// https://github.com/akkadotnet/Akka.Streams.Kafka/issues/415 from prematurely terminating consumers.
+        /// </summary>
+        private IImmutableList<ConsumeResult<K, V>> _unRequestedMessages = ImmutableList<ConsumeResult<K, V>>.Empty;
+
         /// <summary>
         /// Keeps commit senders that need a reply once stashed commits are made.
         /// </summary>
         private IImmutableList<IActorRef> _rebalanceCommitSenders = ImmutableArray<IActorRef>.Empty;
 
-        private ImmutableList<TopicPartition> _pausedPartitions = ImmutableList<TopicPartition>.Empty;
+        private ImmutableList<TopicPartition> _revokedPartitions = ImmutableList<TopicPartition>.Empty;
```

Comment:
Trying to solve this issue via pausing partitions regularly ain't gonna work, so instead we're going to use partition revocation information to help keep our queue of `_unRequestedMessages` sanitized.